core/retries/src/main/java/software/amazon/awssdk/retries/internal/StandardRetryStrategyImpl.java

```diff
@@ -0,0 +1,356 @@
+    private final TokenBucketStore tokenBucketStore;
+
+    private StandardRetryStrategyImpl(Builder builder) {
+        this.predicates = Collections.unmodifiableList(Validate.paramNotNull(builder.predicates, "predicates"));
+        this.maxAttempts = Validate.isPositive(builder.maxAttempts, "maxAttempts");
+        this.circuitBreakerEnabled = builder.circuitBreakerEnabled;
+        this.backoffStrategy = Validate.paramNotNull(builder.backoffStrategy, "backoffStrategy");
+        this.exceptionCost = builder.exceptionCost;
+        this.tokenBucketStore = Validate.paramNotNull(builder.tokenBucketStore, "tokenBucketStore");
+    }
+
+    @Override
+    public AcquireInitialTokenResponse acquireInitialToken(AcquireInitialTokenRequest request) {
+        logAcquireInitialToken(request);
+        return AcquireInitialTokenResponseImpl.create(
+            DefaultRetryToken.builder().scope(request.scope()).build(), Duration.ZERO);
+    }
+
+    @Override
+    public RefreshRetryTokenResponse refreshRetryToken(RefreshRetryTokenRequest request) {
+        DefaultRetryToken token = asStandardRetryToken(request.token());
+        AcquireResponse acquireResponse = requestAcquireCapacity(request, token);
+
+        // Check if we meet the preconditions needed for retrying. These will throw if the expected condition is not meet.
+        // 1) is retryable?
+        throwOnNonRetryableException(request, acquireResponse);
+        // 2) max attempts reached?
+        throwOnMaxAttemptsReached(request, acquireResponse);
+        // 3) can we acquire a token?
+        throwOnAcquisitionFailure(request, acquireResponse);
+
+        // Refresh the retry token and compute the backoff delay.
+        DefaultRetryToken refreshedToken = refreshToken(request, acquireResponse);
+        Duration backoff = backoffStrategy.computeDelay(refreshedToken.attempt());
+
+        // Take the max delay between the suggested delay and the backoff delay.
+        Duration suggested = request.suggestedDelay().orElse(Duration.ZERO);
+        Duration finalDelay = maxOf(suggested, backoff);
+
+        logRefreshTokenSuccess(refreshedToken, acquireResponse, finalDelay);
+        return RefreshRetryTokenResponseImpl.create(refreshedToken, finalDelay);
+    }
+
+    @Override
+    public RecordSuccessResponse recordSuccess(RecordSuccessRequest request) {
+        DefaultRetryToken token = asStandardRetryToken(request.token());
+
+        // Update the circuit breaker token bucket.
+        ReleaseResponse releaseResponse = updateCircuitBreakerTokenBucket(token);
+
+        // Refresh the retry token and return
+        DefaultRetryToken refreshedToken = refreshRetryTokenAfterSuccess(token, releaseResponse);
+
+        // Log success and return.
+        logRecordSuccess(token, releaseResponse);
+        return RecordSuccessResponse.create(refreshedToken);
+    }
+
+    @Override
+    @ToBuilderIgnoreField({"DEFAULT_EXCEPTION_TOKEN_COST", "DEFAULT_TOKEN_BUCKET_SIZE"})
```

Comment:
This can be removed after [#3933](https://github.com/aws/aws-sdk-java-v2/pull/3933) gets merged.